### DISCUSSION TITLE
How can I identify if a mesh is periodic from the basis name?

### DISCUSSION THREAD

**Discussion by JustinPrivitera** [mfem/mfem#4556] [Q&A] [answered]
Hello, VisIt developer here. Is there a way to tell if an mfem mesh is periodic from the grid function basis name?

These are periodic: `L2_T1_3D_P1`, `L2_T1_2D_P1`
These are not: `H1_3D_P2`, `L2_2D_P3`, `L2_T2_3D_P1`

Is it the T1? The P1?

**Comment by pazner** [comment]
All periodic MFEM meshes will use L2 spaces (so the prefix will be `L2`), but not all meshes with L2 spaces will be periodic. I don't know of any way to tell if a mesh is actually periodic (in any case, what about a mesh that is periodic only in one dimension vs. two dimensions, etc.?).

**Comment by JustinPrivitera** [comment]
This is how we test in VisIt if something is periodic:

```c++
    if (mesh && 
        mesh->GetNodes() && 
        mesh->GetNodes()->FESpace() &&
        mesh->GetNodes()->FESpace()->FEColl() &&
        dynamic_cast<const L2_FECollection *>(mesh->GetNodes()->FESpace()->FEColl()))
    {
        // Check if the mesh is periodic.
        AVT_MFEM_INFO("High Order Mesh is periodic; falling back to Legacy LOR.");
        return LegacyRefineMeshToVTK(mesh, domain, lod);
    }
```

**Reply by JustinPrivitera** [reply]
We would like to be able to tell when we are looking at the metadata, long before we load meshes for specific domains.

**Reply by JustinPrivitera** [reply]
Metadata has basis names for grid functions.

**Reply by pazner** [reply]
Not directly relevant to your question, but the check above can be shortened to

```c++
if (mesh && mesh->GetNodalFESpace() && mesh->GetNodalFESpace()->IsDGSpace()) { … }
```

**Reply by JustinPrivitera** [reply]
It is written that way to avoid segfaults.

**Reply by pazner** [reply]
Using the metadata to do the equivalent check to what you have written is just checking if the prefix is `L2`.

**Reply by pazner** [reply]
Yes, I understand. The version I wrote will not segfault.

**Reply by JustinPrivitera** [reply]
Ah I see; I did not read closely. Thank you.

**Reply by JustinPrivitera** [reply]
> Using the metadata to do the equivalent check to what you have written is just checking if the prefix is `L2`.

Like you said before, L2 does not necessarily mean periodic. So we need more information than just whether or not something is L2.

**Reply by pazner** [reply]
I don't know if there is any other way, but if the space is L2, then the test you posted above will consider the mesh periodic even if it's not "really periodic". And so checking if the prefix is `L2` will be equivalent to that check.

If the space is L2, the mesh is "potentially periodic" — I don't know of a way in MFEM, using metadata or otherwise, to identify if a "potentially periodic" mesh is "actually periodic".

**Reply by JustinPrivitera** [reply]
> I don't know if there is any other way, but if the space is L2, then the test you posted above will consider the mesh periodic even if it's not "really periodic". And so checking if the prefix is `L2` will be equivalent to that check.

So, how can we know if we can use `mfem::Mesh::MakeRefined()` or not?

**Reply by pazner** [reply]
`Mesh::MakeRefined` should work for periodic meshes. What goes wrong?

**Reply by JustinPrivitera** [reply]
Legacy approach:
![image](https://github.com/user-attachments/assets/3e86a95e-83be-4d0b-8c3a-f1cf69547776)

MakeRefined:
![image](https://github.com/user-attachments/assets/13009108-222f-4d0d-8c3c-345fe62872be)

**Reply by pazner** [reply]
```c++
#include "mfem.hpp"

using namespace mfem;

int main(int argc, char *argv[])
{
   // 1. Parse command-line options.
   const char *mesh_file = "../data/periodic-hexagon.mesh";
   int order = 4;
   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh", "Mesh file");
   args.AddOption(&order, "-o", "--order", "Order");
   args.ParseCheck();
   Mesh mesh(mesh_file);
   Mesh mesh_lor = Mesh::MakeRefined(mesh, order, BasisType::ClosedUniform);
   std::ofstream f("mesh_lor.mesh");
   mesh_lor.Print(f);
   return 0;
}
```

Original:
![mesh_orig](https://github.com/user-attachments/assets/2487ae73-7359-4671-865e-a35770c94bda)

Refined:
![mesh_refined](https://github.com/user-attachments/assets/0dbd3a43-0e27-4efa-b7c4-8f78dc57b7e4)

**Reply by JustinPrivitera** [reply]
I'm not seeing that behavior with my periodic hexagon, even using `BasisType::ClosedUniform`. Can you give me your periodic hexagon on LC or otherwise point me towards that example? My LC username is `justin`.

**Reply by pazner** [reply]
I am using this file: https://github.com/mfem/mfem/blob/master/data/periodic-hexagon.mesh

**Reply by JustinPrivitera** [reply]
Using MakeRefined:
![image](https://github.com/user-attachments/assets/fcdb27e0-bbf7-4227-aa84-f49a73690a37)

**Reply by JustinPrivitera** [reply]
We are on mfem4.6, are there changes since 4.6 that affect this?

**Reply by pazner** [reply]
I don't think so. There was #4414, but I don't think it explains the behavior you are seeing. Does the code snippet I pasted above work for you (e.g. using GLVis to look at the mesh file)?

**Reply by pazner** [reply]
Are you at some point deleting the mesh nodes (for example by calling `Mesh::SetCurvature(0)`)? Is VisIt drawing the mesh using the node coordinates or vertex coordinates? I can see how you would get output like what you are getting if you go by vertex coordinates rather than node coordinates.

**Reply by JustinPrivitera** [reply]
We don't need to support periodicity - our goal is to disconnect the mesh. I think MakeRefined returns a periodic mesh when you give it a periodic mesh. After the LOR, we need the final points in space to build our LOR mesh outside of MFEM.

We've discussed this before with MFEM team members and determined that we need a special implementation (probably in MFEM) that handles the periodic case.

**Reply by cyrush** [reply]
We do use the vertex coordinates. In other cases of make refined those are valid.

But for the periodic case were faced with how to transform the nodes into physical space, we don't have a recipe for that.
We last tocuhed base on this in 2022 - at the time there wasn't a good solution, so we reverted to use our old LOR for these cases. 

It sounds like we can't detect periodic without an operation the examines the actual topology, and not sure even that case is clear.

**Reply by pazner** [reply]
Yes, `MakeRefined` will return a periodic mesh if you give it a periodic mesh. The points in space are given by the nodal points. I don't think it will be so easy to "disconnect" the mesh (i.e. to take a periodic mesh and make it non-periodic).

**Reply by pazner** [reply]
I think the best option for detecting periodicity is just to check if the space is L2, which may have false positives, but will never have false negatives.

If you just need the physical coordinates of the nodes, that is easy to get with MFEM (even for periodic meshes).

What seems to make this tricky is that MFEM handles periodic meshes by combining/merging vertices. For example, a periodic line segment actually only has one endpoint vertex, rather than two endpoint vertices. So if you are working directly with the mesh vertices, things will get tangled like you see above.

**Reply by JustinPrivitera** [reply]
We discussed an algorithm to do exactly that back in 2022. Tzanio said this:

> For the case of LOR periodic mesh, I can imagine an algorithm like this:
> - initialize vertex coordinates to nans
> - for each element:
> - use the element nodes to get vertex coordinates
> - for each vertex:
> - if the previous value was nan or round-off close to the new one, set to new one
> - otherwise, this is one of the periodic vertices: add a new vertex with the new coordinates, and use that in the element

We are using our legacy approach for now which duplicates all the vertices.

That is beside the point; we are looking for a way to detect periodicity early so we can set up metadata in VisIt properly.

**Reply by JustinPrivitera** [reply]
Do you know what T1 or P1 in the basis name refers to?

**Reply by pazner** [reply, marked as answer]
The algorithm that Tzanio describes sounds about right.

The meaning of the names is described here:

https://github.com/mfem/mfem/blob/17955e114020af340e9a06a66ebef43e05012d9c/fem/fe_coll.hpp#L128-L129

If the space has the form `L2_dD_Pp`, then the space is $d$ dimensional with degree $p$ polynomials.

If the space has the form `L2_Tt_dD_Pp`, then it additionally uses basis type `t`, where the basis types are defined here:

https://github.com/mfem/mfem/blob/17955e114020af340e9a06a66ebef43e05012d9c/fem/fe/fe_base.hpp#L28-L42

**Reply by JustinPrivitera** [reply]
Thank you, this is helpful. I think for now our approach will be to use the legacy refinement method for meshes with grid function bases that are L2.

**Comment by JustinPrivitera** [comment]
Your function `mfem::Mesh::MakeRefined()` cannot handle periodic meshes, so we use a legacy path. We want to know whether or not we can use `mfem::Mesh::MakeRefined()` down the line, and we want to know that when we are processing metadata.

**Comment by JustinPrivitera** [comment]
I think we've discussed this before with @tzanio and @v-dobrev. Are either of you aware of how to check if a mesh is periodic?
